### PR TITLE
Fix emitting events for pods preempted by Kueue [pod groups]

### PR DESCRIPTION
**Pull request by mimowo**
#### What type of PR is this?

/kind cleanup
/kind bug

#### What this PR does / why we need it:

Solve two issues with events for pod groups:
1. the event indicating the a pod in pod group was emitted repeatedly, while the intention is to sent it once
2. the event was emitted only for the first pod in group, but because the Load function may return the pods in different order, we could get both pods covered with at least one event

This PR also adds e2e test to ensure the events are emitted per pod in the group.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

Fix for unreleased feature
```release-note
NONE
```

### CODE REVIEW

**Comment by mimowo** [comment]
/test pull-kueue-test-e2e-main-1-27
MultiKueue failure

File: test/e2e/singlecluster/pod_test.go [hunk 1]
```diff
@@ -273,6 +275,16 @@ var _ = ginkgo.Describe("Pod groups", func() {
 		})
 
 		ginkgo.It("should allow to preempt the lower priority group", func() {
+			podList := corev1.PodList{}
+			podWatcher, err := k8sClient.Watch(ctx, &podList, &client.ListOptions{
```
Comment (alculquicondor) [inline test/e2e/singlecluster/pod_test.go:279]: The watcher should be for Events, not for Pods.

Comment (mimowo) [reply test/e2e/singlecluster/pod_test.go:279]: I see, let me try that.

Comment (mimowo) [reply test/e2e/singlecluster/pod_test.go]: Done, PTAL

**Comment by alculquicondor** [comment]
Please check the failure

**Comment by mimowo** [comment]
for the record: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_kueue/1667/pull-kueue-test-e2e-main-1-29/1752753221249536000

**Comment by mimowo** [comment]
I reproduce this failure locally. What is more it seems these events are really not always sent. Locally I run the `kubectl get events -npod-e2e` (fixed the namespace for ease) and these are the events:

```
DELETED    37s         Warning   ErrWorkloadCompose   pod/default-priority-group-0                   'default-priority-group' group total count is less than the actual number of pods in the cluster
DELETED    37s         Normal    CreatedWorkload      pod/default-priority-group-0                   Created Workload: pod-e2e/default-priority-group
DELETED    37s         Normal    Started              pod/default-priority-group-0                   Admitted by clusterQueue cq
DELETED    37s         Normal    Scheduled            pod/default-priority-group-0                   Successfully assigned pod-e2e/default-priority-group-0 to kind-worker
DELETED    35s         Normal    Stopped              pod/default-priority-group-0                   Preempted to accommodate a higher priority Workload
DELETED    37s         Normal    Pulled               pod/default-priority-group-0                   Container image "gcr.io/k8s-staging-perf-tests/sleep:v0.1.0" already present on machine
DELETED    37s         Normal    Created              pod/default-priority-group-0                   Created container c
DELETED    36s         Normal    Started              pod/default-priority-group-0                   Started container c
DELETED    37s         Normal    Started              pod/default-priority-group-1                   Admitted by clusterQueue cq
DELETED    37s         Normal    Scheduled            pod/default-priority-group-1                   Successfully assigned pod-e2e/default-priority-group-1 to kind-worker
DELETED    37s         Normal    Pulled               pod/default-priority-group-1                   Container image "gcr.io/k8s-staging-perf-tests/sleep:v0.1.0" already present on machine
DELETED    37s         Normal    Created              pod/default-priority-group-1                   Created container c
DELETED    36s         Normal    Started              pod/default-priority-group-1                   Started container c
```
Note that there is no `Preempted to accommodate a higher priority Workload` for `pod/default-priority-group-1`. 

Also, I checked with another command that both pods are actually preempted (I get the `TerminationTarget` for both of them).

I will investigate tomorrow why is this so. In the meanwhile, let me know @alculquicondor if you have some insights.

**Comment by alculquicondor** [comment]
I would first question whether Kueue is somehow skipping recording an event for certain pods.... or we are just facing the nature of events: writing events is not something that the client retries.

If the former, then we can fix it. If the later, I guess it's worse than I thought and we can only write reliable unit test, but not integration tests. Although I seem to remember we have reliable integration tests for events in kube-scheduler (under the assumption that event recording failures are highly uncommon).

Can you try to verify that?

**Comment by alculquicondor** [comment]
Yup, we have reliable integration tests for events in kube-scheduler https://github.com/kubernetes/kubernetes/blob/57e9981dc2745b2f99d01bf0df6bedb0f16ed051/test/integration/scheduler/scheduler_test.go#L286

So we need to check the reconciler code that somehow isn't skipping events for particular pods.

**Comment by alculquicondor** [comment]
We can probably accept a certain level of flakiness for this test, but it should be less than 1%

File: test/e2e/singlecluster/pod_test.go [hunk 2]
```diff
@@ -326,6 +341,24 @@ var _ = ginkgo.Describe("Pod groups", func() {
...
+			ginkgo.By("Use events to observe the default-priority pods are getting preempted", func() {
+				preemptedPods := sets.New[types.NamespacedName]()
+				gomega.Eventually(func(g gomega.Gomega) sets.Set[types.NamespacedName] {
+					select {
+					case evt, ok := <-eventWatcher.ResultChan():
+						gomega.Expect(ok).To(gomega.BeTrue())
+						event, ok := evt.Object.(*v1.Event)
+						gomega.Expect(ok).To(gomega.BeTrue())
+						objKey := types.NamespacedName{Namespace: event.InvolvedObject.Namespace, Name: event.InvolvedObject.Name}
+						if defaultGroupPods.Has(objKey) && event.Reason == "Stopped" && event.Message == "Preempted to accommodate a higher priority Workload" {
```
Comment (alculquicondor) [inline test/e2e/singlecluster/pod_test.go:353]: let's not compare the message.

**Review by alculquicondor — COMMENTED** [review]
The test itself looks about right
/approve

**Comment by mimowo** [comment]
> I would first question whether Kueue is somehow skipping recording an event for certain pods.... or we are just facing the nature of events: writing events is not something that the client retries.

Kueue is skipping sending of the events. I added temporarily log line around https://github.com/kubernetes-sigs/kueue/blob/a8c11f351858cd159a3f823b15860918c0490cbc/pkg/controller/jobframework/reconciler.go#L748, and can confirm this is ocasionally not called for ` pod/default-priority-group-1`. It is always called for ` pod/default-priority-group-0`. Looking for the solution.

**Comment by mimowo** [comment]
Ok, I understand the root cause. We assume with the stop interface that the only stopped Job is represented with the object, but this is not the case for composable jobs, where multiple pods are stopped.

Also, I know why it sometimes passes, this is because the `stopJob` routine is called multiple times before the workload finished (around 10 times), and the `Load` function returns the pods in a non-deterministic order. So sometimes we get the event for `pod-1`.

There is another issue, we actually send the `Stopped` now event multiple times for the same pod, because the Stop function returns always true, assuming no errors: https://github.com/kubernetes-sigs/kueue/blob/a8c11f351858cd159a3f823b15860918c0490cbc/pkg/controller/jobs/pod/pod_controller.go#L462.

I think one solution would be to extend the API of Stop by returning a slice of actually stopped Jobs "now", or slice of extra jobs stopped now. WDYT?

File: test/e2e/singlecluster/pod_test.go (hunk 2 above)
Comment (mimowo) [reply test/e2e/singlecluster/pod_test.go]: done

**Comment by mimowo** [comment]
> We can probably accept a certain level of flakiness for this test, but it should be less than 1%

After the fix it is quite stable, looped it locally and passed >200 times in a row .

File: pkg/controller/jobs/job/job_controller.go [hunk 3]
```diff
@@ -157,14 +157,14 @@ func (j *Job) Suspend() {
...
+func (j *Job) Stop(ctx context.Context, c client.Client, podSetsInfo []podset.PodSetInfo, _ jobframework.StopReason, eventMsg string) ([]client.Object, error) {
+	var stoppedNow []client.Object
 	if !j.IsSuspended() {
 		j.Suspend()
 		if err := c.Update(ctx, j.Object()); err != nil {
-			return false, fmt.Errorf("suspend: %w", err)
+			return stoppedNow, fmt.Errorf("suspend: %w", err)
 		}
-		stoppedNow = true
+		stoppedNow = []client.Object{j.Object()}
```
Comment (yaroslava-serdiuk) [inline pkg/controller/jobs/job/job_controller.go:167]: Why do we need slice if we return 0 or 1 object?

Comment (mimowo) [reply pkg/controller/jobs/job/job_controller.go:167]: In the current approach I implement the `Stop` function from the `JobWithCustomStop` interface. 

This interface is also used by pod groups, where we want to emit the events about multiple objects (pods). 

I guess we could have the `Stop` function under `ComposableJob`. WDYT? cc @alculquicondor

Comment (mimowo) [reply pkg/controller/jobs/job/job_controller.go]: I just pushed a commit which reverts the change for the Job controller, and introduces a Stop function in the ComposableJob. I think it looks better, PTAL.

File: pkg/controller/jobs/pod/pod_controller_test.go [hunk 4]
```diff
@@ -2451,6 +2764,9 @@ func TestReconciler(t *testing.T) {
 			if diff := cmp.Diff(tc.wantWorkloads, gotWorkloads.Items, tc.workloadCmpOpts...); diff != "" {
 				t.Errorf("Workloads after reconcile (-want,+got):\n%s", diff)
 			}
+			if diff := cmp.Diff(tc.wantEvents, recorder.RecordedEvents, cmp.Transformer("sortedEvents", utiltesting.SortedEvents)); diff != "" {
```
Comment (alculquicondor) [inline pkg/controller/jobs/pod/pod_controller_test.go:2767]: There is a specific opt for sorting :) https://pkg.go.dev/github.com/google/go-cmp/cmp/cmpopts#SortSlices

**Review by alculquicondor — COMMENTED** [review]
/lgtm
/approve

File: pkg/controller/jobs/pod/pod_controller_test.go (hunk 4 above)
Comment (mimowo) [reply pkg/controller/jobs/pod/pod_controller_test.go:2767]: Indeed, opened a small cleanup https://github.com/kubernetes-sigs/kueue/pull/1691

**Pull request merged**